docs: create keptn metrics getting started (#1375)
keptn/lifecycle-toolkit: docs/content/en/docs/getting-started/metrics/_index.md

Whether a rollout is good, whether to scale up or down.
Your observability data may come
from multiple observability solutions --
Datadog, Dynatrace, Lightstep, Honeycomb, Splunk,
or data directly from your cloud provider such as AWS, Google, or Azure.

The Keptn Lifecycle Toolkit hooks directly into Kubernetes primitives
so minimal configuration is required.

The
[Kubernetes metric server](https://github.com/kubernetes-sigs/metrics-server)
requires that you maintain point-to-point integrations
from Argo Rollouts, Flux, KEDA, and HPA.
Each has plugins but it is difficult to maintain them,
especially if you are using multiple tools,
and multible observability platforms,
and multiple instance of some tools or observability platforms.
The Keptn Metrics Server unifies and standardizes access to this data.

## Using this exercise

This exercise is based on the
[simplenode-dev](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd)
example.
You can clone that repo to access it locally
or just look at it for examples
as you implement the functionality "from scratch"
on your local Kubernetes deployment cluster.

The steps to implement metrics in an existing cluster are:

> 1. [Install the Keptn Lifecycle Toolkit](../../install/install)

Reviewer: Most likely this will make the link checker happy
```suggestion
1. [Install the Keptn Lifecycle Toolkit](../../install/install/)
```

### View available metrics

Use the following command to view
the metrics that are configured in your cluster.
This example displays the two metrics we configured above:

```shell
kubectl get KeptnMetrics -A
```

```shell
NAMESPACE       NAME              PROVIDER       QUERY
simplenode-dev  availability-slo  dev-dynatrace  func:slo.availability_simplenodeservice
simplenode-dev  available-cpus    dev-prometheus sum(kube_node_status_capacity{resource=`cpu`})
```

## Run the metrics

As soon as you define your `KeptnMetricsProvider` and `KeptnMetric` resources,
the Lifecycle Toolkit begins collecting the metrics you defined.
You do not need to do anything else.

## Observing the metrics

The metrics can be retrieved
through CRs and through the Kubernetes Metric API.

The syntax to retrieve metrics from the CR is:

```shell
kubectl get keptnmetrics.metrics.keptn.sh -n <namespace> <metric-name>
```

For example, the output for the `available-cpus` metric looks like:

```shell
$ kubectl get keptnmetrics.metrics.keptn.sh -n simplenode-dev available-cpus

NAME             PROVIDER     QUERY                                           VALUE

> cpu-throttling   my-provider  sum(kube_node_status_capacity{resource=`coy})   6.000

Reviewer: ```suggestion
cpu-throttling   my-provider  sum(kube_node_status_capacity{resource=`cpu`})   6.000
```

defines the information you want to gather,
specified as a query for the particular observability platform
you are using.
You can define any type of metric from any data source.

In our example, we define two bits of information to retrieve:

- Number of CPUs, derived from the `dev-prometheus` data platform
- `availability` SLO, derived from the `dev-dynatrace` data platform

Each of these are configured to fetch data every 10 seconds
but you could configure a different `fetchIntervalSeconds` value
for each metric.

The
[keptn-metric.yaml](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/simplenode-dev/keptn-metric.yaml)
file for our example looks like:

```yaml
apiVersion: metrics.keptn.sh/v1alpha2
kind: Keptnmetric
metadata:
  name: available-cpus
  namespace: simplenode-dev
spec:
  provider:
    name: dev-prometheus
  query: "sum(kube_node_status_cvapacity{resources`cpu`})"
  fetchIntervalSeconds: 10
...
apiVersion: metrics.keptn.sh/v1alpha2
kind: Keptnmetric
metadata:
  name: availability-slo
  namespace: simplenode-dev
spec:
  provider:
    name: dev-dynatrace
  query: "func:slo.availability_simplenodeservice"
  fetchIntervalSeconds: 10

> ...

Reviewer 2: to stay consistent with the yaml syntax, we should use `---` to separate the two KeptmMetric objects - this way this can be directly copied into a file and be applied

```yaml
kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta2/namespaces/<namespace>/keptnmetrics.metrics.sh/<metric-name>/<metric-name>"
```

For example, the output for the `available-cpus` looks like:

```yaml
$ kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta2/namespaces/simplenode-dev/keptnmetrics.metrics.sh/available-cpus/available-cpus"

{
  "kind": "MetricValueList",
  "apiVersion": "custom.metrics.k8s.io/v1beta2",
  "metadata": {},
  "items": [
    {
      "describedObject": {
        "kind": "KeptnMetric",
        "namespace": "simplenode-dev",
        "name": "available-cpus",
        "apiVersion": "metrics.keptn.sh/v1alpha2"
      },
      "metric": {
        "name": "available-cpus",
        "selector": {}
      },
      "timestamp": "2023-05-11T08:05:36Z",
      "value": "6"
    }
  ]
}
```

You can also display the graphics using a dashboard such as Grafana.

## Implementing autoscaling with HPA

The Kubernetes HorizontalPodAutoscaler (HPA)
uses metrics to provide autoscaling for the cluster.
HPA can retrieve KeptnMetrics and use it to implement HPA.
See Using the [HorizontalPodAutoscaler](../../implmenting/evaluatemetrics)

> for detailed information.

Reviewer: I think we can get rid of these lines since the next paragraph is a duplicate that has a better link :)

Author: done manually

file for our example looks like:

```yaml
apiVersion: metrics.keptn.sh/v1alpha2
kind: Keptnmetric
metadata:
  name: available-cpus
  namespace: simplenode-dev
spec:
  provider:
    name: dev-prometheus
  query: "sum(kube_node_status_cvapacity{resources`cpu`})"
  fetchIntervalSeconds: 10
...
apiVersion: metrics.keptn.sh/v1alpha2
kind: Keptnmetric
metadata:
  name: availability-slo
  namespace: simplenode-dev
spec:
  provider:
    name: dev-dynatrace
  query: "func:slo.availability_simplenodeservice"
  fetchIntervalSeconds: 10
...
```

Note the following:

- Populate one YAML file per metric
  then apply all of them.
- Each metric is assigned a unique `name`.
- The value of the `spec.provider.name` field
  must correspond to the name assigned in a
  the `metadata.name` field of a `KeptnMetricsProvider` resource.
- Information is fetched in on a continuous basis
at a rate specified by the value of the `spec.fetchIntervalSeconds` field.

### View available metrics

Use the following command to view
the metrics that are configured in your cluster.
This example displays the two metrics we configured above:

> TODO: Is the syntax of the output changed
> to include the `name` of the provider (`dev-prometheus` or `dev-dynatrace`?

Reviewer: I think we can remove the todo now
```suggestion
```

or data directly from your cloud provider such as AWS, Google, or Azure.

The
[Kubernetes metric server](https://github.com/kubernetes-sigs/metrics-server)
requires that you maintain point-to-point integrations
from Argo Rollouts, Flux, KEDA, and HPA.
Each has plugins but it is difficult to maintain them,
especially if you are using multiple tools,
and multible observability platforms,
and multiple instance of some tools or observability platforms.
The Keptn Metrics Server unifies and standardizes access to this data.

## Using this exercise

This exercise is based on the
[simplenode-dev](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd)
example.
You can clone that repo to access it locally
or just look at it for examples
as you implement the functionality "from scratch"
on your local Kubernetes deployment cluster.

The steps to implement metrics are:

1. Install and configure Keptn Lifecycle Toolkit
   - [Bring or create a Kubernetes cluster](#bring-or-create-a-kubernetes-deployment-cluster)
   - [Install the Keptn Lifecycle Toolkit on your cluster](#install-klt-on-your-cluster)
   - [Enable KLT for your cluster](#enable-klt-for-your-cluster)

>    - [Integrate KLT with your cluster](#integrate-klt-with-your-deployment)

Author: This list was pretty much there and tweaked in https://github.com/keptn/lifecycle-toolkit/pull/1400 .  Text here rewritten to reference that list.

This line tells KLT to handle the namespace

## Integrate KLT with your deployment

To integrate KLT with your deployment, annotate the Kubernetes
[Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
resource.
In this example, this is defined in the
[simplenode-dev-deployment.yaml](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/simplenode-dev/simplenode-dev-deployment.yaml)
file, which includes the following lines:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: simplenode
  namespace: simplenode-dev
...
template:
    metadata:
      labels:
        app: simplenode
        app.kubernetes.io/name: simplenodeservice
      annotations:
        # keptn.sh/app: simpleapp
        keptn.sh/workload: simplenode
        keptn.sh/version: 1.0.2
        keptn.sh/pre-deployment-evaluations: evaluate-dependencies
        keptn.sh/pre-deployment-tasks: notify
        keptn.sh/post-deployment-evaluations: evaluate-deployment
        keptn.sh/post-deployment-tasks: notify
...
```

For more information about using annotations and labels
to integrate KLT into your deployment cluster, see

> [Integrate KLT with your applications](../../implementing/integrate/_index.md).

Author: It is done!

file for our example looks like:

```yaml
apiVersion: metrics.keptn.sh/v1alpha2
kind: Keptnmetric
metadata:
  name: available-cpus
  namespace: simplenode-dev
spec:
  provider:
    name: dev-prometheus
  query: "sum(kube_node_status_cvapacity{resources`cpu`})"
  fetchIntervalSeconds: 10
...
apiVersion: metrics.keptn.sh/v1alpha2
kind: Keptnmetric
metadata:
  name: availability-slo
  namespace: simplenode-dev
spec:
  provider:
    name: dev-dynatrace
  query: "func:slo.availability_simplenodeservice"
  fetchIntervalSeconds: 10
...
```

Note the following:

* Populate one YAML file per metric
  then apply all of them.
* Each metric is assigned a unique `name`.
* The value of the `spec.provider.name` field
  must correspond to the name assigned in a
  the `metadata.name` field of a `KeptnMetricsProvider` resource.
* Information is fetched in on a continuous basis
at a rate specified by the value of the `spec.fetchIntervalSeconds` field.

### View available metrics

Use the following command to view
the metrics that are configured in your cluster.
This example displays the two metrics we configured above:

TODO: Is the syntax of the output changed

> to include the `name` of the provider (`dev-prometheus` or `dev-dynatrace`?

Author: Thanks!

```shell
get KeptnMetrics -A
```

```shell
NAMESPACE       NAME              PROVIDER   QUERY
simplenode-dev  availability-slo  dynatrace  func:slo.availability_simplenodeservice
simplenode-dev  available-cpus    prometheus sum(kube_node_status_capacity{resource=`coy})
```

## Run the metrics

TODO: Do I need to start and stop anything to start gathering metrics
or could I theoretically just put these pieces into my cluster
and would it start gathering metrics that I could then view?

## Observing the metrics

TODO: Do we want to say anything about running these metrics,

> viewing the results, perhaps from CLI and from Grafana?

Reviewer: Here I would show how we can retrieve the metrics through CRDs and through the Kubernetes Metric API.

From CRDs the syntax is `kubectl get keptnmetrics.metrics.keptn.sh -n <namespace> <metric-name>`. Here's an example with the output for the `available-cpus`

```bash
$ kubectl get keptnmetrics.metrics.keptn.sh -n simplenode-dev available-cpus
NAME             PROVIDER     QUERY                                           VALUE
cpu-throttling   my-provider  sum(kube_node_status_capacity{resource=`coy})   6.000
```

For the kubernetes APIs, the syntax is the following `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta2/namespaces/<namespace>/keptnmetrics.metrics.sh/<metric-name>/<metric-name>"`. Here's an example with the output for the `available-cpus`

```bash
$ kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta2/namespaces/simplenode-dev/keptnmetrics.metrics.sh/available-cpus/available-cpus"
{
  "kind": "MetricValueList",
  "apiVersion": "custom.metrics.k8s.io/v1beta2",
  "metadata": {},
  "items": [
    {
      "describedObject": {
        "kind": "KeptnMetric",
        "namespace": "simplenode-dev",
        "name": "available-cpus",
        "apiVersion": "metrics.keptn.sh/v1alpha2"
      },
      "metric": {
        "name": "available-cpus",
        "selector": {}
      },
      "timestamp": "2023-05-11T08:05:36Z",
      "value": "6"
    }
  ]
}
```

Author: Wonderful!  It's in!

## Implementing autoscaling with HPA

The Kubernetes HorizontalPodAutoscaler (HPA)
uses metrics to provide autoscaling for the cluster.
HPA can retrieve KeptnMetrics and use it to implement HPA.

> See Using the HorizontalPodAutoscaler](../../implmenting/evaluatemetrics)

Reviewer: ```suggestion
See Using the [HorizontalPodAutoscaler](../../implmenting/evaluatemetrics)
```

weight: 25
---

The Keptn metrics component of the Keptn Lifecycle Toolkit
allow you to define any type of metric
from multiple instances of any type of data source in your Kubernetes cluster.

> You may have tools like Argo, Flux, KEDA, HBA, Keptn

Reviewer: ```suggestion
You may have tools like Argo, Flux, KEDA, HPA, or Keptn
```

that need observability data to make automated decisions.
Whether a rollout is good, whether to scale up or down.
Your observability data may come
from multiple observability solutions --

> Datadog, Dynatrace, data in AWS, Google, and Azure --
> and include data in Lifestep (?) and Honeycomb or Splunk.

Reviewer: ```suggestion
Datadog, Dynatrace, Lightstep, Honeycomb, Splunk, or data directly from your cloud provider such as AWS, Google, and Azure.
```

## Using this exercise

This exercise is based on the
[simplenode-dev](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd)
example.
You can clone that repo to access it locally
or just look at it for examples
as you implement the functionality "from scratch"
on your local Kubernetes deployment cluster.

The steps to implement pre- and post-deployment orchestration are:

1. [Bring or create a Kubernetes cluster](#bring-or-create-a-kubernetes-deployment-cluster)
1. [Install the Keptn Lifecycle Toolkit on your cluster](#install-klt-on-your-cluster)
1. [Enable KLT for your cluster](#enable-klt-for-your-cluster)
1. [Integrate KLT with your cluster](#integrate-klt-with-your-cluster)

> 1. Define metrics to use

Reviewer: To avoid "define" multiple times
```suggestion
1. Configure metrics to use
```

   * [View available metrics](#view-available-metrics)

See the
[Introducing Keptn Lifecycle Toolkit](https://youtu.be/449HAFYkUlY)
video for a demonstration of this exercise.

## Bring or create a Kubernetes deployment cluster

You can run this exercise on an existing Kubernetes cluster
or you can create a new cluster.
For personal study and demonstrations,
this exercise runs well on a local Kubernetes cluster.
See [Bring or Install a Kubernetes Cluster](../../install/k8s.md).

## Install KLT on your cluster

Install the Keptn Lifecycle Toolkit on your cluster
by executing the following command sequence:

```shell
helm repo add klt https://charts.lifecycle.keptn.sh
helm repo update
helm upgrade --install keptn klt/klt \
   -n keptn-lifecycle-toolkit-system --create-namespace --wait
```

If you only want to use Keptn's metrics features,
you can install just the `metrics-operator`
bu modifying Helm values.
See
[Install KLT](../../install/install.md)
for more information about installing the Lifecycle Toolkit.

To verify that the `metrics-operator` is installed in your cluster,
run the following command:

```shell
kubectl get pods -n keptn-lifecycle-toolkit-system
```

> The output shows all components that are running on your system.

Reviewer: I would create a snippet and cross-reference it from here and the install page. This way, we don't need to maintain the instruction on how to install KLT in multiple places. You can see an example how to include snippets here: https://github.com/thisthat/lifecycle-controller/blob/c5c45558795d99a7789c48b9940d336990f46d3e/docs/content/en/docs/getting-started/generic-gs/_index.md?plain=1#L43

## Enable KLT for your cluster

To enable KLT for your cluster, annotate the Kubernetes
[Namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/)
resource.
In this example, this is defined in the
[simplenode-dev-ns.yaml](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/simplenode-dev/simplenode-dev-ns.yaml)
file, which looks like this:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: simplenode-dev
  annotations:
    keptn.sh/lifecycle-toolkit: "enabled"
```

You see the annotation line that enables `lifecycle-toolkit`.

> This line tells the webhook to handle the namespace

Reviewer: ```suggestion
This line tells KLT to handle the namespace
```

> ## Integrate KLT with your cluster

Reviewer: ```suggestion
## Integrate KLT with your deployment
```

> To integrate KLT with your cluster, annotate the Kubernetes

Reviewer: ```suggestion
To integrate KLT with your deployment, annotate the Kubernetes
```

      annotations:
        # keptn.sh/app: simpleapp
        keptn.sh/workload: simplenode
        keptn.sh/version: 1.0.2
        keptn.sh/pre-deployment-evaluations: evaluate-dependencies
        keptn.sh/pre-deployment-tasks: notify
        keptn.sh/post-deployment-evaluations: evaluate-deployment
        keptn.sh/post-deployment-tasks: notify
...
```

For more information about using annotations and labels
to integrate KLT into your deployment cluster, see
[Integrate KLT with your applications](../../implementing/integrate.md).

The [Kubernetes metric server](https://github.com/kubernetes-sigs/metrics-server)
requires that you maintain point-to-point integrations
from Argo Rollouts, Flux, KEDA, and HPA.
Each has plugins but it is difficult to maintain them,
especially if you are using multiple tools
and multible observability platforms.
The Keptn Metrics Server unifies and standardizes access to this data.

The steps to implement Keptn metrics are:

## Install and configure KLT

Use the Helm Chart to install the Keptn Metrics Server
as part of the Lifecycle Toolkit
or completely stand-alone.
 See
[Install KLT using the Helm Chart](../../install/install.md/#use-helm-chart).

> -- End of Probably goes --

Reviewer: This was already provided few section before

Author: Whoops -- that "End of Probably goes -- marker indicates that I meant to delete that.  I'll fix.

Author: Fixed

* `dev-dynatrace`

You can specify a virtually unlimited number of providers,
including multiple instances of each observability platform.
Each one must be assigned a unique name,
identified by the type of platform it is
and the URL.

> Note: The video and example application use an older syntax
  of the `KeptnMetricsProvider` and `KeptnMetric` resources.
  The syntax shown in this document is correct for v0.7.1 and later.

Definition of
[dev-prometheus](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/simplenode-dev/keptn-prometheus-provider.yaml)
data source:

```yaml
kind: KeptnMetricsProvider
metadata:
  name: dev-prometheus
  namespace: simplenode-dev
spec:
  type: prometheus
  targetserver: "http://prometheus-k8s-monitoring-svc.cluster.local:9090"
```

Definition of the
[dev-dynatrace](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/simplenode-dev/dynatrace-provider.yaml.tmp)
data source.
Note that the `dev-dynatrace` server is protected by a secret key
so that information is included in the provider definition:

```yaml
kind: KeptnMetricsProvider
metadata:
  name: dev-dynatrace
  namespace: simplenode-dev
spec:
  type: dynatrace

>   targetServer: "https://hci34192.live.dynatrace.com

Reviewer: ```suggestion
  targetServer: "https://hci34192.live.dynatrace.com"
```

  secretKeyRef
    name: dynatrace
    key: DT_TOKEN
...
```

> 

Reviewer: The secret key contains the API token to authenticate with your Dynatrace tenant. You can provide your own token by creating a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/) like the following:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: dynatrace
  namespace: simplenode-dev
type: Opaque
data:
  DT_TOKEN: "<api-token>"
```

Author: I don't think this belongs in a Getting Started Guide but it belongs somewhere -- then the GSG can link to that other doc for people who want to learn about it.  We have a bit of different information about secrets that is currently squirreled away on the KeptnTaskDefinition (along with info about context) -- it doesn't belong there but that is where it was in the old docs so I left it for now.  I created https://github.com/keptn/lifecycle-toolkit/issues/1390 so we remember to do it.

### Define KeptnMetric information

The [KeptnMetric](../../yaml-crd-ref/metric.md) resource
defines the information you want to gather,
specified as a query for the particular observability platform
you are using.
You can define any type of metric from any data source.

In our example, we define two bits of information to retrieve:

* Number of CPUs, derived from the `dev-prometheus` data platform
* `availability` SLO, derived from the `dev-dynatrace` data platform

Each of these are configured to fetch data every 10 seconds
but you could configure a different `fetchIntervalSeconds` value
for each metric.

The
[keptn-metric.yaml](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/simplenode-dev/keptn-metric.yaml)
file for our example looks like:

```yaml
apiVersion: metrics.keptn.sh/v1alpha2
kind: Keptnmetric
metadata:
  name: available-cpus
  namespace: simplenode-dev
spec:
  provider:
    name: dev-prometheus

>   query: "sum(kube_node_status_cvapacity{resources`cpu`})

Reviewer: ```suggestion
  query: "sum(kube_node_status_cvapacity{resources`cpu`})"
```

  fetchIntervalSeconds" 10
...
apiVersion: metrics.keptn.sh/v1alpha2
kind: Keptnmetric
metadata:
  name: availability-slo
  namespace: simplenode-dev
spec:
  provider:
    name: dev-dynatrace
  query: "func:slo.availability_simplenodeservice"
  fetchIntervalSeconds" 10
...
```

Note the following:

* You populate one YAML file

> that includes all the metrics for your cluster.

Reviewer: I suggest to don't add this. Users can provide one metric per yaml file and apply all of them.

Author: Ah, then I will write it that way.  Actually, we should define this for all the yaml files for the different CRDs.  For example, for KeptnConfig, apparently the preference is to have a single yaml file that defines both configurations.  But we'll have on metric per yaml and I believe one provider per yaml.  What about evaluationdefinition and taskdefinition -- one per yaml or one yaml for everything or some mix?   This doesn't need to be specified in Getting Started but on the reference pages, then we need to write the Getting Started guide to match the recommendations.  And then get Andi to follow the recommendations in his videos, too ;-)

Author: This now reads: "Populate one YAML file per metric then apply all of them."
